feat: Preview icon themes

joshuamegnauth54 opened the pull request:
(Maybe) closes: #208 

![icon_picker](https://github.com/pop-os/cosmic-settings/assets/48846352/11cde602-0bb3-4b5a-9af7-49dd8d133efd)

Edit: Actually it may be a little different than #208 but it still seems pretty useful.

mmstick commented:
Requires design approval. @pop-os/ux

git-f0x commented:
Personally, it looks a bit busy when always visible like that.
Maybe have icon theme selection in a separate window (similar to the Add Applet window), and in the place of the current drop-down for icon selection, maybe have a clickable name of the current theme, which opens the icon theme selection window.

maria-komarova commented:
The features have been added before the design was finalized. The design is described here now: [#235](https://github.com/pop-os/cosmic-settings/issues/235). Please let me know if there are any questions.

joshuamegnauth54 commented:
Thank you for your comments, everyone! I'll try to implement these changes.

maria-komarova commented:
Thanks for working on this!

joshuamegnauth54 commented:
![theme_prev01](https://github.com/pop-os/cosmic-settings/assets/48846352/ac7e14fb-bc5a-4700-8173-bab70c538e1c)
![theme_prev02](https://github.com/pop-os/cosmic-settings/assets/48846352/c5b2f29b-8bd8-4ef4-aab0-5925f9a4fd75)

I tested it with a few random themes that I placed into a fake `XDG_DATA_DIRS` to make sure it works.

mmstick commented on cosmic-settings/src/pages/desktop/appearance.rs:
```diff
@@ -439,6 +446,45 @@ impl Page {
+        let theme = cosmic::theme::active();
+        let theme = theme.cosmic();
+        cosmic::iced::widget::column![
+            // Export theme choice to GNOME
```
This option will apply to other toolkits as well. Not just GTK/GNOME.

mmstick commented on cosmic-settings/src/pages/desktop/appearance.rs:
```diff
@@ -1571,3 +1631,102 @@ async fn set_gnome_icon_theme(theme: String) {
+    cosmic::iced::widget::column![
+        button(
+            cosmic::iced::widget::column![
+                row::Row::new()
```
use `cosmic::widget::row()` instead.

mmstick commented on cosmic-settings/src/pages/desktop/appearance.rs:
```diff
@@ -1571,3 +1631,102 @@ async fn set_gnome_icon_theme(theme: String) {
+) -> Element<'static, Message> {
+    let theme = cosmic::theme::active();
+    let theme = theme.cosmic();
+    cosmic::iced::widget::column![
```
Use `cosmic::widget::column()` instead.

mmstick commented on cosmic-settings/src/pages/desktop/appearance.rs:
```diff
@@ -829,6 +876,13 @@ impl Page {
+                self.context_view = Some(ContextView::Experimental);
+
+                cosmic::command::message(crate::app::Message::OpenContextDrawer(
+                    fl!("experimental").into(),
```
From the designs, this should be an empty string

maria-komarova commented:
Looks great!

Just a few minor things base on the screenshot above (sorry if those were already fixed):
- The copy on the list item on the main content area should say "Experimental settings".
- There should be no header in the Context Drawer or if there is one, then is should say "Experimental Settings".
- The symbolic icon at the end of the list item on the main content area should be "go-next-symbolic" (actions icons).
- "Icon theme" inside Context drawer should be using "Header" text style and it should be spaced from the following thumbnails less (should be @space_xxs). The spacing between "Icon theme" and thumbnails should be less than the spacing between the list item above and "Icon theme" to maintain content hierarchy. 
- We might also want to cut the theme name after two lines. I know there is a text wrapping issue, it's been mentioned in other places. But still better have consistent spacing between rows of thumbnails. We can cut the long names with "...". 

Are icon theme thumbnails using image buttons? Or the same widget wallpapers do? It looks like it has a different corner radius and no additional selection indicator in the left bottom corner.

maria-komarova commented:
Looks great!

Just a few minor things base on the screenshot above (sorry if those were already fixed):

- The copy on the list item on the main content area should say "Experimental settings".
- There should be no header in the Context Drawer or if there is one, then is should say "Experimental Settings".
- The symbolic icon at the end of the list item on the main content area should be "go-next-symbolic" (actions icons).
- "Icon theme" inside Context drawer should be using "Header" text style and it should be spaced from the following thumbnails less (should be @space_xxs). The spacing between "Icon theme" and thumbnails should be less than the spacing between the list item above and "Icon theme" to maintain content hierarchy.
- We might also want to cut the theme name after two lines. I know there is a text wrapping issue, it's been mentioned in other places. But still better have consistent spacing between rows of thumbnails. We can cut the long names with "...".

Are icon theme thumbnails using image buttons? Or the same widget wallpapers do? It looks like it has a different corner radius and no additional selection indicator in the left bottom corner.

joshuamegnauth54 commented:
![menu](https://github.com/pop-os/cosmic-settings/assets/48846352/f3310363-323b-487e-a31e-9a99e315d1ac)
![context](https://github.com/pop-os/cosmic-settings/assets/48846352/2a1cb385-a02d-4317-8e1f-f98f84da75d1)

Here is an update to check if I'm on the right track.

To do tomorrow:
* Header text style
* Shortening long icon titles
* I think I have to pad the buttons a bit

TheESN commented:
BTW It seems like this screen utilizes the smallest icon size variations from themes. Can you make it use bigger variations?

mmstick commented on cosmic-settings/src/pages/desktop/appearance.rs:
```diff
@@ -1571,3 +1631,131 @@ async fn set_gnome_icon_theme(theme: String) {
+
+    cosmic::widget::column()
+        .push(
+            button::Button::new_image(
```
This should be `cosmic::widget::button::image()`

mmstick commented on cosmic-settings/src/pages/desktop/appearance.rs:
```diff
@@ -1571,3 +1631,131 @@ async fn set_gnome_icon_theme(theme: String) {
+    cosmic::widget::column()
+        .push(
+            button::Button::new_image(
+                cosmic::iced::widget::column![
```
`cosmic::widget::column()`, `cosmic::widget::column::with_capacity()`, or `cosmic::widget::column::with_children()`.

mmstick commented on cosmic-settings/src/pages/desktop/appearance.rs:
```diff
@@ -1571,3 +1631,131 @@ async fn set_gnome_icon_theme(theme: String) {
+        .push(
+            button::Button::new_image(
+                cosmic::iced::widget::column![
+                    cosmic::widget::row::row()
```
`cosmic::widget::row()`

mmstick commented on cosmic-settings/src/pages/desktop/appearance.rs:
```diff
@@ -1571,3 +1631,131 @@ async fn set_gnome_icon_theme(theme: String) {
+                                .map(|handle| handle.icon().size(ICON_PREV_SIZE))
+                        )
+                        .spacing(theme.space_xs()),
+                    row::Row::new()
```
`cosmic::widget::row()`

mmstick commented on cosmic-settings/src/pages/desktop/appearance.rs:
```diff
@@ -439,6 +446,48 @@ impl Page {
+            // Icon theme previews
+            // cosmic::iced::widget::column![text(&*ICON_THEME), text(&*ICON_THEME_DESC).size(10)]
+            //     .spacing(2),
+            cosmic::iced::widget::column![
```
`cosmic::widget::column()`, `cosmic::widget::column::with_capacity()`, or `cosmic::widget::column::with_children()`

joshuamegnauth54 commented:
Hello fellow cosmonauts!

Changes for today:
* Truncate long theme names
* Use [heading](https://pop-os.github.io/libcosmic/cosmic/widget/text/fn.heading.html) for "Icon theme"
* I padded the buttons a bit to fit the design better

![new01](https://github.com/pop-os/cosmic-settings/assets/48846352/d1cf951b-8380-4a3b-bc73-8b887fda5fef)

@TheESN I'm using `32` for the size since it seems like a good default, but I agree that some flexibility will be nice! Maybe a future patch can add a slider to increase or decrease the size?

I can also change the default for now too. I'm aesthetically challenged, so I'm not sure which size is best.

## Icon size 48
![size_48](https://github.com/pop-os/cosmic-settings/assets/48846352/da71dfcd-c2b9-40bc-ba1f-bb69e3914f88)

## Icon size 64
![size_64](https://github.com/pop-os/cosmic-settings/assets/48846352/81def67f-e4b3-4be4-abb5-955ccf53e738)

## Icon size 128
:joy_cat: 
![size_128](https://github.com/pop-os/cosmic-settings/assets/48846352/d3993ad8-e88f-415a-b65b-33ccd5440b31)

32 (current) and 48 both look pretty nice.

TheESN commented:
> Hello fellow cosmonauts!
> 
> Changes for today:
> 
>     * Truncate long theme names
> 
>     * Use [heading](https://pop-os.github.io/libcosmic/cosmic/widget/text/fn.heading.html) for "Icon theme"
> 
>     * I padded the buttons a bit to fit the design better
> 
> 
> ![new01](https://private-user-images.githubusercontent.com/48846352/323858829-d1cf951b-8380-4a3b-bc73-8b887fda5fef.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM1MTI5MjgsIm5iZiI6MTcxMzUxMjYyOCwicGF0aCI6Ii80ODg0NjM1Mi8zMjM4NTg4MjktZDFjZjk1MWItODM4MC00YTNiLWJjNzMtOGI4ODdmZGE1ZmVmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE5VDA3NDM0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMxMjRiODkyNGM3ZmQxZDc5YzE1NjIzOTdmNTI1ZTkxZDUyOGQ5NTNiNGI5ZTFjMjRjNDYxZjIyZWU3MTJmOTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.qbHbSHbFUCGAALiIP4WcRaqof7sG-RKGMoVYyTT7NHw)
> 
> @TheESN I'm using `32` for the size since it seems like a good default, but I agree that some flexibility will be nice! Maybe a future patch can add a slider to increase or decrease the size?
> 
> I can also change the default for now too. I'm aesthetically challenged, so I'm not sure which size is best.
> ## Icon size 48
> 
> ![size_48](https://private-user-images.githubusercontent.com/48846352/323861273-da71dfcd-c2b9-40bc-ba1f-bb69e3914f88.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM1MTI5MjgsIm5iZiI6MTcxMzUxMjYyOCwicGF0aCI6Ii80ODg0NjM1Mi8zMjM4NjEyNzMtZGE3MWRmY2QtYzJiOS00MGJjLWJhMWYtYmI2OWUzOTE0Zjg4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE5VDA3NDM0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRlZGVlOTNlYWU3MjAwOWVjMDdmOTg2ZjA2OGM4ZmY5OTk0ZDFkOWY5YTZmMWU4Y2RiOTQwYzE3NDEwZjE4NGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.LCtTgBBk4J5e2lZk1g-qpmMUzh7N_cjq3yZ5wTM3GKI)
> ## Icon size 64
> 
> ![size_64](https://private-user-images.githubusercontent.com/48846352/323861342-81def67f-e4b3-4be4-abb5-955ccf53e738.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM1MTI5MjgsIm5iZiI6MTcxMzUxMjYyOCwicGF0aCI6Ii80ODg0NjM1Mi8zMjM4NjEzNDItODFkZWY2N2YtZTRiMy00YmU0LWFiYjUtOTU1Y2NmNTNlNzM4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE5VDA3NDM0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ1MDhkZTA0YWM0OTI2MmVlMzVkNjBmMWZiYzI2MDc2NjUwMmFjN2NhYjdjNjNkZDU4OTdmYzYwMzFlNmZmYjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.vz0Igj6vzO2AfHtOctNlCwFGO7RRNri9oUNWEAGcKk0)
> ## Icon size 128
> 
> 😹 ![size_128](https://private-user-images.githubusercontent.com/48846352/323861638-d3993ad8-e88f-415a-b65b-33ccd5440b31.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM1MTI5MjgsIm5iZiI6MTcxMzUxMjYyOCwicGF0aCI6Ii80ODg0NjM1Mi8zMjM4NjE2MzgtZDM5OTNhZDgtZTg4Zi00MTVhLWI2NWItMzNjY2Q1NDQwYjMxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE5VDA3NDM0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMzODdkNjVhNzQ1OWY4MzM2NzIxZWU4NmZjYWI2ZGZjMDU5YjM0YWY2N2RjZmFkZTY0NTdkN2ViMzhhYTg3MTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.lQ2RGD87_uDOPpU-SftC9eXiXlMe6S66ogns7QPM0Fg)
> 
> 32 (current) and 48 both look pretty nice.

I didn't mean the actual size, but their style. Icon themes usually specify several icons for different sizes, so smaller icons have less details, and bigger ones have more details. 

From screenshots that you sent I see that it uses the smallest size of icons and just scales them without changing to more detailed variations.

For reference that's how 32x32 Pop folder should look like
![изображение](https://github.com/pop-os/cosmic-settings/assets/102585253/f7debbcd-cb5d-4eb0-a2d3-e02282f66f08)

And that's how 48x48 should look like
![изображение](https://github.com/pop-os/cosmic-settings/assets/102585253/9ffe1230-7881-45df-986d-6b51a709e9fc)

The one that your preview is using is actually 16x16
![изображение](https://github.com/pop-os/cosmic-settings/assets/102585253/3d539cb8-e2bb-45c6-b0d4-0360fee74dd5)

joshuamegnauth54 commented:
> I didn't mean the actual size, but their style. Icon themes usually specify several icons for different sizes, so smaller icons have less details, and bigger ones have more details.
> 
> From screenshots that you sent I see that it uses the smallest size of icons and just scales them without changing to more detailed variations.

Alright. I think I fixed it now:

![icon_size_fix](https://github.com/pop-os/cosmic-settings/assets/48846352/cedaea56-c57e-4d05-b4fb-557bd3ad7f01)

It turns out that I was evaluating the sizes too late. I updated the code to check for a few different sizes before falling back to the default too.

The situation for `La Capitaine` and `Luv` is different. I'm unsure if the problem rests with how I manually installed the themes or with `libcosmic`. Basically, the theme name and the directory have to match. `La Capitaine`'s `index.theme` lists the name as `Name=La Capitaine` but the directory is `la-capitaine`. I renamed the directory to match `index.theme` which worked.

The [spec](https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html) states:
> Each theme is stored as subdirectories of the base directories. The internal name of the theme is the name of the subdirectory, although the user-visible name as specified by the theme may be different. Hence, theme names are case sensitive, and are limited to ASCII characters. Theme names may also not contain comma or space.

The spec also lists an example of a theme titled `Birch` which is located in a directory called `birch`. I _think_ my issue with `La Capitaine` and `Luv` may be a bug?

TheESN commented:
Looks definitely better! Though now I am wondering why adwaita is missing settings icon in preview

joshuamegnauth54 commented:
![symbolic_fallback](https://github.com/pop-os/cosmic-settings/assets/48846352/eda5a2cc-8a47-4e2b-88de-f26ba589d35a)

Adwaita doesn't have `preferences-system` but it does have a symbolic iteration of that icon. I updated the code to try the icons listed in the design or fallback to the symbolic variant if the requested icon isn't found.

Alsooo, Vibrancy-Dark was missing icons because I didn't copy over the themes it inherited to my test `XDG_DATA_DIR`. I fixed it to ensure my code still works.

TheESN commented:
> Adwaita doesn't have preferences-system but it does have a symbolic iteration of that icon

Well now I am wondering how is this icon called then?

![изображение](https://github.com/pop-os/cosmic-settings/assets/102585253/9799be79-cbed-4029-8d9c-9c420d6734a7)

joshuamegnauth54 commented:
I couldn't find an icon that looks like that in `Adwaita`. It's possible that that's an app icon stored in `hicolors` instead of `Adwaita` itself which would allow non-`Adwaita` themes to display the correct icon. `hicolors` is the default fallback for every theme which means it's a useful place to store app icons.

TheESN commented:
> I couldn't find an icon that looks like that in `Adwaita`. It's possible that that's an app icon stored in `hicolors` instead of `Adwaita` itself which would allow non-`Adwaita` themes to display the correct icon. `hicolors` is the default fallback for every theme which means it's a useful place to store app icons.

Well maybe it's GNOME-only icon then. Kinda weird that they decided to go this way, so symbolic fallback should be fine then

mmstick commented:
Is this ready for merging?

maria-komarova commented:
Looks ready to me.
